src/Npgsql/NpgsqlConnector.cs

```diff
@@ -936,10 +916,68 @@ await socket.ConnectAsync(endpoint)
                     if (i == endpoints.Length - 1)
                         throw new NpgsqlException($"Failed to connect to {endpoint}", e);
                 }
+            }
+
+            Task<IPAddress[]> GetHostAddressesAsync(NpgsqlTimeout timeout, CancellationToken cancellationToken)
+            {
+                // .NET 6.0 added cancellation support to GetHostAddressesAsync, which allows us to implement real
+                // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
+                // and raise the exception, but the actual task is left running.
+
+#if NET6_0_OR_GREATER
+                CancellationTokenSource? combinedCts = null;
+                try
+                {
+                    var finalCt = cancellationToken;
+
+                    if (timeout.IsSet)
+                    {
+                        combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                        combinedCts.CancelAfter((int)timeout.CheckAndGetTimeLeft().TotalMilliseconds);
+                        finalCt = combinedCts.Token;
+                    }
+
+                    return Dns.GetHostAddressesAsync(Host, cancellationToken);
```

Comment:
dotnet/runtime#33420 added the support for cancellable dns lookup only for windows. I do not thinks it's really worth it to check whether we're on windows or not, so let's leave it as is.

Reply:
@vonzshik but the point is that this change removes the WithCancellationAndTimeout hack for net6.0. This means that on Linux (and only with net6.0!), a networking issue will cause the DNS lookup to block forever, where with earlier versions of Npgsql the call would at least timeout/cancel (even if the actual DNS operation continues blocking). That seems like it could be quite a regression.

Reply:
Oh, I've meant to leave it as it's already implemented in main, not in this pr, sorry :bread:

Reply:
Ah I see :)

@manandre can you please do this change?

Reply:
Done.

Reply:
See my comment.